<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">


# Introduction to Pandas



Pandas is the most popular Python package for managing data sets. It's used extensively by data scientists.

### Learning Objectives

- Define the anatomy of DataFrames.
- Explore data with DataFrames.

### Lesson Guide

- [Introduction to `pandas`](#introduction)
- [Loading CSV Files](#loading_csvs)
- [Exploring Your Data](#exploring_data)
- [Data Dimensions](#data_dimensions)
- [DataFrames vs. Series](#dataframe_series)
- [Using the `.info()` Function](#info)
- [Using the `.describe()` Function](#describe)
- [Independent Practice](#independent_practice)
- [Pandas Indexing](#indexing)
- [Creating DataFrames](#creating_dataframes)
- [Checking Data Types](#dtypes)
- [Renaming and Assignment](#renaming_assignment)
- [Logical Filtering](#filtering)
- [Sorting](#sorting)
- [Review](#review)

<a id='introduction'></a>

### What is a dataframe?

---
The concept of a "dataframe" comes from the world of statistical software used in empirical research; 
- Generally refers to "tabular" data: a data structure representing cases (rows), each of which consists of a number of observations or measurements (columns)
- **Each row** is treated as a **single observation** of **multiple "variables"** 
- The row ("record") datatype can be **heterogenous** (a tuple of different types) 
- The column datatype must be **homogenous**. 
- Data frames usually contain some **metadata** in addition to **data**; for example, column and row names (unlike Numpy by default)

<a id='introduction'></a>

### What is `pandas`?

---

- A data analysis library — **P**anel **D**ata **S**ystem.
- It was created by Wes McKinney and open sourced by AQR Capital Management, LLC in 2009.
- It's implemented in highly optimized Python/Cython.
- It's the **most ubiquitous tool** used to start data analysis projects within the Python scientific ecosystem.


### Pandas Use Cases

---

- Cleaning data/munging.
- Exploratory analysis.
- Structuring data for plots or tabular display.
- Joining disparate sources.
- Modeling.
- Filtering, extracting, or transforming. 


![](https://snag.gy/tpiLCH.jpg)

![](https://snag.gy/1V0Ol4.jpg)

### Common Outputs

---

With `pandas` you can:

- Export to databases
- Integrate with `matplotlib`
- Collaborate in common formats (plus a variety of others)
- Integrate with Python built-ins (**and `numpy`!**)


### Importing `pandas`

---

Import `pandas` at the top of your notebook like so:

In [1]:
import pandas as pd
import numpy as np

Recall that the **`import pandas as pd`** syntax nicknames the `pandas` module as **`pd`** for convenience.

<a id='loading_csvs'></a>

### Loading a CSV into a DataFrame

---

`pandas` can load many types of files, but one of the most commonly used for storing data is a ```.csv```. As an example, let's load a data set on diamond characteristics from the ```./datasets``` directory:

In [2]:
diamonds = pd.read_csv('./datasets/diamonds.csv')

In [3]:
diamonds

carat        cut color clarity  depth  table  price     x     y     z
0       0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1       0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2       0.23       Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3       0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4       0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75
...      ...        ...   ...     ...    ...    ...    ...   ...   ...   ...
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64

[53940 rows x 10 columns]

This creates a `pandas` object called a **DataFrame**. DataFrames are powerful containers, featuring many built-in functions for exploring and manipulating data.

We will barely scratch the surface of DataFrame functionality in this lesson, but, throughout this course, you will become an expert at using them.

In short, a dataframe is a supercharged 2D array:
    - it has the data
    - it has information about it (meta-data - like columns names, etc...)

<a id='exploring_data'></a>

### Exploring Data using DataFrames

---

DataFrames come with built-in functionality that makes data exploration easy. 

To start, let's look at the **"header"** of your data using the ```.head()``` function. If run alone in a notebook cell, it will show you the first handful of columns in the data set, along with the first five rows.

In [4]:
diamonds.head(10)

carat        cut color clarity  depth  table  price     x     y     z
0   0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23       Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75
5   0.24  Very Good     J    VVS2   62.8   57.0    336  3.94  3.96  2.48
6   0.24  Very Good     I    VVS1   62.3   57.0    336  3.95  3.98  2.47
7   0.26  Very Good     H     SI1   61.9   55.0    337  4.07  4.11  2.53
8   0.22       Fair     E     VS2   65.1   61.0    337  3.87  3.78  2.49
9   0.23  Very Good     H     VS1   59.4   61.0    338  4.00  4.05  2.39

If we want to see the last part of our data, we can use the ```.tail()``` function equivalently.

In [5]:
diamonds.tail(10)

carat        cut color clarity  depth  table  price     x     y     z
53930   0.71    Premium     E     SI1   60.5   55.0   2756  5.79  5.74  3.49
53931   0.71    Premium     F     SI1   59.8   62.0   2756  5.74  5.73  3.43
53932   0.70  Very Good     E     VS2   60.5   59.0   2757  5.71  5.76  3.47
53933   0.70  Very Good     E     VS2   61.2   59.0   2757  5.69  5.72  3.49
53934   0.72    Premium     D     SI1   62.7   59.0   2757  5.69  5.73  3.58
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64

In [6]:
diamonds[50000:50010]

carat        cut color clarity  depth  table  price     x     y     z
50000   0.57      Ideal     F     VS1   61.0   56.0   2193  5.36  5.40  3.28
50001   0.76       Fair     H     SI1   65.5   59.0   2193  5.75  5.66  3.74
50002   0.74    Premium     I     VS1   62.9   57.0   2193  5.79  5.76  3.63
50003   0.74       Fair     H     VS1   66.3   63.0   2193  5.63  5.50  3.69
50004   0.70    Premium     E     SI1   62.4   60.0   2194  5.72  5.63  3.54
50005   0.71      Ideal     G     SI2   59.5   57.0   2194  5.86  5.78  3.46
50006   0.70  Very Good     F     SI2   60.7   58.0   2195  5.73  5.77  3.49
50007   0.83       Good     J     SI1   63.8   58.0   2195  5.95  5.97  3.80
50008   0.57      Ideal     H      IF   62.2   56.0   2195  5.30  5.33  3.30
50009   0.51       Good     F    VVS2   62.4   63.0   2195  5.05  5.08  3.16

<a id='data_dimensions'></a>

### Data Dimensions

---

It's always good to look at the dimensions of your data. The ```.shape``` property will tell you how many rows and columns are contained within your DataFrame.

In [7]:
diamonds.shape

(53940, 10)

As you can see, we have 53940 rows and 10 columns.

You'll also notice that this function operates the same as `.shape` for `numpy` arrays/matricies. **`pandas` makes use of **numpy** under its hood** for optimization and speed.

You can look up the names of your columns using the ```.columns``` property.


In [8]:
diamonds.columns

Index(['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'price', 'x', 'y',
       'z'],
      dtype='object')

Accessing a specific column is easy. You can use **bracket** syntax just like you would with **Python dictionaries**, using the column's string name to extract it.

In [9]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [10]:
diamonds['carat'].head()

0    0.23
1    0.21
2    0.23
3    0.29
4    0.31
Name: carat, dtype: float64

As you can see, we can also use the ```.head()``` function on a single column, which is represented as a `pandas` Series object.

With a **list of strings**, you can also access a column (as a DataFrame instead of a Series).

In [11]:
diamonds[['carat']].head()

carat
0   0.23
1   0.21
2   0.23
3   0.29
4   0.31

In [12]:
diamonds[['cut','carat']].head()

cut  carat
0    Ideal   0.23
1  Premium   0.21
2     Good   0.23
3  Premium   0.29
4     Good   0.31

<a id='dataframe_series'></a>

### DataFrame vs. Series

---

There is an important difference between using a list of strings versus only using a string with a column's name: When you use a list containing the string, it returns another **DataFrame**. But, when you only use the string, it returns a `pandas` **Series** object.

In [13]:
print(type(diamonds['cut']))

<class 'pandas.core.series.Series'>


In [14]:
print(type(diamonds[['cut']]))

<class 'pandas.core.frame.DataFrame'>


**Breakout (2min):** What's the difference between `pandas` Series and DataFrame objects?

In [15]:
# diamonds.price%.head() cannot because got percentage !!!

As long as your column names **don't contain any spaces** or other specialized characters (underscores are OK), you can access a column as a property of a DataFrame.  

**Get in the habit of referencing your Series columns using `df['my_column']` rather than with object notation (`df.my_column`)**. There are many edge cases in which the object notation does not work, along with nuances as to how `pandas` will behave.

In [16]:
diamonds['cut'].head()

0      Ideal
1    Premium
2       Good
3    Premium
4       Good
Name: cut, dtype: object

In [17]:
diamonds[['cut']].head()

cut
0    Ideal
1  Premium
2     Good
3  Premium
4     Good

Remember: This will be a **Series** object, not a **DataFrame**.

<a id='info'></a>

### Examining Your Data With `.info()`

---

When getting acquainted with a new data set, `.info()` should be **the first thing** you examine.

**Types** are very important. They affect the way data will be **represented** in our machine learning models, how data can be joined, whether or not math operators can be applied, and instances in which you can encounter unexpected results.

> _Typical problems that arise when working with new data sets include_:
> - Missing values.
> - Unexpected types (string/object instead of int/float).
> - Dirty data (commas, dollar signs, unexpected characters, etc.).
> - Blank values that are actually "non-null" or single white-space characters.

`.info()` is a function available on every **DataFrame** object. It provides information about:

- The name of the column/variable attribute.
- The type of index (RangeIndex is default).
- The count of non-null values by column/attribute.
- The type of data contained in the column/attribute.
- The unique counts of **dtypes** (`pandas` data types).
- The memory usage of our data set.

#### For example: 

In [18]:
diamonds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    53940 non-null  float64
 1   cut      53940 non-null  object 
 2   color    53940 non-null  object 
 3   clarity  53940 non-null  object 
 4   depth    53940 non-null  float64
 5   table    53940 non-null  float64
 6   price    53940 non-null  int64  
 7   x        53940 non-null  float64
 8   y        53940 non-null  float64
 9   z        53940 non-null  float64
dtypes: float64(6), int64(1), object(3)
memory usage: 4.1+ MB


<a id='describe'></a>

### Summarizing Data with `.describe()`

---

The ```.describe()``` function is useful for taking a quick look at your data. It returns some basic descriptive statistics.

For our example, use the ```.describe()``` function on only the ```carat``` column.

In [19]:
diamonds['carat'].describe()

count    53940.000000
mean         0.797940
std          0.474011
min          0.200000
25%          0.400000
50%          0.700000
75%          1.040000
max          5.010000
Name: carat, dtype: float64

You can also use it on multiple columns, such as ```carat``` and ```price```. They will need to be numeric types. 

In [20]:
diamonds[['carat','price']].describe()

carat         price
count  53940.000000  53940.000000
mean       0.797940   3932.799722
std        0.474011   3989.439738
min        0.200000    326.000000
25%        0.400000    950.000000
50%        0.700000   2401.000000
75%        1.040000   5324.250000
max        5.010000  18823.000000

```.describe()``` gives us the following statistics:

- **Count**, which is equivalent to the number of cells (rows).
- **Mean**, or, the average of the values in the column.
- **Std**, which is the standard deviation.
- **Min**, a.k.a., the minimum value.
- **25%**, or, the 25th percentile of the values.
- **50%**, or, the 50th percentile of the values ( which is the equivalent to the median).
- **75%**, or, the 75th percentile of the values.
- **Max**, which is the maximum value.

<img src="https://snag.gy/AH6E8I.jpg">

#### Summary Functions

There are also built-in math functions that will work on a column of a DataFrame. 

For example, I can use the ```.mean()``` function on the `carat` column to get the mean carat weight for all the diamonds.

In [21]:
diamonds[['carat']].mean()

carat    0.79794
dtype: float64

We can also use them on multiple columns:

In [22]:
diamonds[['carat','price']].max()

carat        5.01
price    18823.00
dtype: float64

if the columns are all numeric you can call them directly on the DataFrame, otherwise you might get a warning:

In [23]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [24]:
diamonds.mean()

C:\Users\tiven\AppData\Local\Temp\ipykernel_11196\2453091466.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  diamonds.mean()


carat       0.797940
depth      61.749405
table      57.457184
price    3932.799722
x           5.731157
y           5.734526
z           3.538734
dtype: float64

<a id='independent_practice'></a>

### Independent Practice

---

Now that we know a little bit about basic DataFrame use, let's practice on a new data set.

> Pro tip: When your cursor is in a string, you can use the "tab" key to browse file system resources and get a relative reference for the files that can be loaded in Jupyter notebook. Remember, you have to use your arrow keys to navigate the files populated in the UI. 

<img src="https://snag.gy/IlLNm9.jpg">

1. Find and load the `cars` data set into a DataFrame (in the `datasets` directory).
2. Print out the columns.
3. What does the data set look like in terms of dimensions?
4. Check the types of each column.
  a. What is the most common type?
  b. How many entries are there?
  c. How much memory does this data set consume?
5. Examine the summary statistics of the data set.

In [25]:
csv_file = "datasets/cars.csv"
cars = pd.read_csv(csv_file)

In [26]:
cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

In [27]:
# column names
cars.columns

Index(['mpg', 'cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear',
       'carb'],
      dtype='object')

In [28]:
# shape
cars.shape

(26, 11)

In [29]:
# column data types, number of entries and memory used. seems like alot of info
#
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26 entries, 0 to 25
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   mpg     26 non-null     float64
 1   cyl     26 non-null     int64  
 2   disp    26 non-null     float64
 3   hp      26 non-null     int64  
 4   drat    26 non-null     float64
 5   wt      26 non-null     float64
 6   qsec    26 non-null     float64
 7   vs      26 non-null     int64  
 8   am      26 non-null     int64  
 9   gear    26 non-null     int64  
 10  carb    26 non-null     int64  
dtypes: float64(5), int64(6)
memory usage: 2.4 KB


In [30]:
# summary stats 
cars.describe()

mpg        cyl        disp          hp       drat         wt  \
count  26.000000  26.000000   26.000000   26.000000  26.000000  26.000000   
mean   19.792308   6.307692  240.373077  138.730769   3.515385   3.346500   
std     6.119930   1.761119  127.182350   59.463977   0.540749   0.993209   
min    10.400000   4.000000   71.100000   52.000000   2.760000   1.615000   
25%    15.275000   4.000000  142.275000   95.500000   3.072500   2.683750   
50%    18.950000   6.000000  241.500000  123.000000   3.460000   3.440000   
75%    22.475000   8.000000  342.000000  180.000000   3.915000   3.767500   
max    33.900000   8.000000  472.000000  245.000000   4.930000   5.424000   

            qsec         vs         am       gear       carb  
count  26.000000  26.000000  26.000000  26.000000  26.000000  
mean   18.244615   0.461538   0.269231   3.423077   2.538462  
std     1.610521   0.508391   0.452344   0.503831   1.240347  
min    15.410000   0.000000   0.000000   3.000000   1.000000  
25%    17.112500   0.000000   0.000000   3.000000   1.250000  
50%    17.990000   0.000000   0.000000   3.000000   2.000000  
75%    19.305000   1.000000   0.750000   4.000000   4.000000  
max    22.900000   1.000000   1.000000   4.000000   4.000000

<a id='indexing'></a>

### `pandas` Indexing 

---

More often than not, we want to operate on or extract specific portions of our data. When we perform indexing on a DataFrame or Series, we can specify a certain section of the data.

`pandas` has three properties you can use for indexing:

- **`.loc`** indexes with the _labels_ for rows and columns.
- **`.iloc`** indexes with the _integer positions_ for rows and columns.

To help clarify these differences, let's first reset the row labels to letters using the ```.index``` attribute:

In [31]:
new_index_values = ['A','B','C','D','E','F','G','H','I','J','K','L','M',
                    'N','O','P','Q','R','S','T','U','V','W','X','Y','Z']
cars.index=new_index_values


In [32]:
cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
A  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
B  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
C  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
D  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
E  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

In [33]:
cars["A":"E"]

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
A  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
B  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
C  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
D  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
E  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

Using the **`.loc`** indexer, we can pull out rows **'B' to 'F'** and the **`hp` and `wt`** columns.

In [34]:
subset = cars.loc[['B','C','D','E','F'], ['hp','wt']]

In [35]:
subset

hp     wt
B  110  2.875
C   93  2.320
D  110  3.215
E  175  3.440
F  105  3.460

In [36]:
subset2 = cars.loc['B':'J', ['hp','wt']]
subset2

hp     wt
B  110  2.875
C   93  2.320
D  110  3.215
E  175  3.440
F  105  3.460
G  245  3.570
H   62  3.190
I   95  3.150
J  123  3.440

We can do the same thing with the **`.iloc`** indexer, but we have to use integers for the location.

In [37]:
subset3 = cars.iloc[1:12, [3,5]]
subset3

hp     wt
B  110  2.875
C   93  2.320
D  110  3.215
E  175  3.440
F  105  3.460
G  245  3.570
H   62  3.190
I   95  3.150
J  123  3.440
K  123  3.440
L  180  4.070

In [38]:
subset = cars.iloc[[1,2,3,4,5], [3,5]]

In [39]:
subset

hp     wt
B  110  2.875
C   93  2.320
D  110  3.215
E  175  3.440
F  105  3.460

If you try to index the rows or columns with integers using **`.loc`**, you will get an error.

Note that you can automatically reorder the data just by reordering the indices you enter when you perform the indexing operation!

While we created an index earlier, we can also use a column to set an index.

In [40]:
cars.index = cars['mpg']

cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
mpg                                                                
21.0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
21.0  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
22.8  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
21.4  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
18.7  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

Is mpg the best feature to use as an index?  

If it isn't we can use the `df.reset_index()` to reset our index.

In [41]:
cars.reset_index(drop=True, inplace=True)
cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

<a id='creating_dataframes'></a>

### Creating DataFrames

---

The simplest way to create your own DataFrame without importing data from a file is to give the ```pd.DataFrame()``` instantiator a dictionary.

In [42]:
mydata = pd.DataFrame({'Letters':['A','B','C'], 'Integers':[1,2,3], 'Floats':[2.2, 3.3, 4.4]})

In [43]:
mydata

Letters  Integers  Floats
0       A         1     2.2
1       B         2     3.3
2       C         3     4.4

As you might expect, the dictionary needs to have lists of values that are all the same length. The keys correspond to the names of the columns, and the values correspond to the data in the columns.

<a id='dtypes'></a>

### Examining Data Types

---

`pandas` comes with a useful property for looking solely at the data types of your DataFrame columns. Use ```.dtypes``` on your DataFrame:

In [44]:
mydata.dtypes

Letters      object
Integers      int64
Floats      float64
dtype: object

This will show you the data type of each column. Strings are stored as a type called "object," as they are not guaranteed to take up a set amount of space (strings can be any length).

<a id='renaming_assignment'></a>

### Renaming and Assignment

---

`pandas` makes it easy to change column names and assign values to your DataFrame.

Say, for example, we want to change the column name `Integers` to `int`:

In [45]:
mydata.rename(columns={mydata.columns[1]:'int'}, inplace=True) # inplace = True updates mydata
print(mydata.columns)

Index(['Letters', 'int', 'Floats'], dtype='object')


In [46]:
mydata

Letters  int  Floats
0       A    1     2.2
1       B    2     3.3
2       C    3     4.4

If you want to change every column name, you can just assign a new list to the ```.columns``` property.

In [47]:
mydata.columns = ['A','B','C']
print(mydata.head())

   A  B    C
0  A  1  2.2
1  B  2  3.3
2  C  3  4.4


<a id='filtering'></a>

### Filtering Logic

---

One of the most powerful features of DataFrames is the ability to use logical commands to filter data.

Subset the ```diamonds``` data for only the rows in which `price` is greater than 7000.

In [48]:
diamonds[diamonds['price'] > 7000]

carat        cut color clarity  depth  table  price     x     y     z
17457   1.01  Very Good     D     VS2   60.6   56.0   7001  6.45  6.51  3.93
17458   1.01  Very Good     F     VS1   63.7   56.0   7001  6.27  6.33  4.01
17459   1.01       Good     F    VVS2   65.9   54.0   7001  6.18  6.27  4.10
17460   1.30    Premium     D     SI2   60.2   59.0   7002  7.00  7.05  4.23
17461   1.00  Very Good     E     VS1   61.2   57.0   7002  6.43  6.47  3.95
...      ...        ...   ...     ...    ...    ...    ...   ...   ...   ...
27745   2.00  Very Good     H     SI1   62.8   57.0  18803  7.95  8.00  5.01
27746   2.07      Ideal     G     SI2   62.5   55.0  18804  8.20  8.13  5.11
27747   1.51      Ideal     G      IF   61.7   55.0  18806  7.37  7.41  4.56
27748   2.00  Very Good     G     SI1   63.5   56.0  18818  7.90  7.97  5.04
27749   2.29    Premium     I     VS2   60.8   60.0  18823  8.50  8.47  5.16

[9273 rows x 10 columns]

#### Filtering on Multiple Conditions

We can also filter on _multiple conditions_. 
The format for multiple conditions is:

`df[ (df['col1'] == value1) & (df['col2'] == value2) ]`

Or, more simply:

`df[ (CONDITION 1) & (CONDITION 2) ]`

Which eventually may evaluate to something like:

`df[ True & False ]`

...on a row-by-row basis. If the end result is `False`, the row is omitted.

_Don't forget parentheses in your conditions!_ This is a common mistake.

#### Example 
Subset the data for `price` greater than 7000 like before, but now, also include where the `cut` is 'Ideal'.

In [49]:
diamonds[(diamonds['price'] > 7000) & (diamonds['cut']== 'Ideal')]

carat    cut color clarity  depth  table  price     x     y     z
17462   1.21  Ideal     I    VVS2   62.7   56.0   7002  6.80  6.83  4.27
17471   1.01  Ideal     E     VS2   61.9   57.0   7013  6.43  6.37  3.96
17472   1.01  Ideal     F     VS2   61.9   54.0   7014  6.45  6.49  4.00
17473   1.04  Ideal     F     VS1   61.4   57.0   7015  6.51  6.56  4.01
17474   1.34  Ideal     I     VS1   62.2   57.0   7016  7.02  7.06  4.38
...      ...    ...   ...     ...    ...    ...    ...   ...   ...   ...
27735   1.60  Ideal     F     VS1   62.0   56.0  18780  7.47  7.52  4.65
27738   2.05  Ideal     G     SI1   61.9   57.0  18787  8.10  8.16  5.03
27741   2.15  Ideal     G     SI2   62.6   54.0  18791  8.29  8.35  5.21
27746   2.07  Ideal     G     SI2   62.5   55.0  18804  8.20  8.13  5.11
27747   1.51  Ideal     G      IF   61.7   55.0  18806  7.37  7.41  4.56

[3266 rows x 10 columns]

In [50]:
# What about diamonds where the cut is Premium OR the carat weight is greater than 0.50?
# "Or" logic - use pipe (|)

In [51]:
diamonds[(diamonds['cut']=='Premium')|(diamonds['carat']>0.5)]

carat        cut color clarity  depth  table  price     x     y     z
1       0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
3       0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
12      0.22    Premium     F     SI1   60.4   61.0    342  3.88  3.84  2.33
14      0.20    Premium     E     SI2   60.2   62.0    345  3.79  3.75  2.27
15      0.32    Premium     E      I1   60.9   58.0    345  4.38  4.42  2.68
...      ...        ...   ...     ...    ...    ...    ...   ...   ...   ...
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64

[39323 rows x 10 columns]

#### Calculations on filtered data

Let's calculate the **mean carat weight** for diamonds with `cut` **Premium**
> Think: What are the component parts of this problem?

In [52]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [53]:
# First find the diamonds where the cut == 'Premium'
# Then select the 'carat' column as a series
# Finally, find the mean
mean_carat = diamonds[diamonds['cut']=='Premium']['carat'].mean()

print(f'{mean_carat:.2f}')

0.89


<a id='sorting'></a>

### Sorting

We can sort the DataFrame by Series, or by the entire DataFrame by specifying which columnd to sort. 

In [54]:
# We can sort individual Series...
cars['mpg'].sort_values().head()

15    10.4
14    10.4
23    13.3
6     14.3
16    14.7
Name: mpg, dtype: float64

In [55]:
# sort in descending order
cars['mpg'].sort_values(ascending=False).head()

19    33.9
17    32.4
18    30.4
25    27.3
7     24.4
Name: mpg, dtype: float64

In [56]:
# Sort the entire DataFrame by the specific column
cars.sort_values('mpg').head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
15  10.4    8  460.0  215  3.00  5.424  17.82   0   0     3     4
14  10.4    8  472.0  205  2.93  5.250  17.98   0   0     3     4
23  13.3    8  350.0  245  3.73  3.840  15.41   0   0     3     4
6   14.3    8  360.0  245  3.21  3.570  15.84   0   0     3     4
16  14.7    8  440.0  230  3.23  5.345  17.42   0   0     3     4

In [57]:
# Or the entire DataFrame by more than one column using lists
# cars.sort_values(by=??? , ascending=??? ).head()

## Independent Practice

With our cars dataset already loaded, let's explore our dataset a bit more thoroughly to gain some familiarity with beginning exploratory analysis.

### 1. Select only data for "cars" when "hp" is "150-200".

In [58]:
cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

In [59]:
cars[ (cars['hp']>=150) & (cars['hp']<=200) ]

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
4   18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2
11  16.4    8  275.8  180  3.07  4.070  17.40   0   0     3     3
12  17.3    8  275.8  180  3.07  3.730  17.60   0   0     3     3
13  15.2    8  275.8  180  3.07  3.780  18.00   0   0     3     3
21  15.5    8  318.0  150  2.76  3.520  16.87   0   0     3     2
22  15.2    8  304.0  150  3.15  3.435  17.30   0   0     3     2
24  19.2    8  400.0  175  3.08  3.845  17.05   0   0     3     2

### 2. Select only rows with index 5-10, for variables / columns "disp" and "hp"

In [60]:
cars.loc[5:10,["disp","hp"]]

disp   hp
5   225.0  105
6   360.0  245
7   146.7   62
8   140.8   95
9   167.6  123
10  167.6  123

### 3. Select the columns by numeric offset 2-5, rows with numeric index 3-7

In [61]:
cars.iloc[ 3:8,2:6]

disp   hp  drat     wt
3  258.0  110  3.08  3.215
4  360.0  175  3.15  3.440
5  225.0  105  2.76  3.460
6  360.0  245  3.21  3.570
7  146.7   62  3.69  3.190

### 4. Find the mean of `hp`


In [62]:
mean_2=cars["hp"].mean()
print(f'{mean_2:.2f}')

138.73


### 4. Find the mean of `hp` for cars with more then 4 cylinders

In [63]:
mean_3=cars[ cars["cyl"] > 4]["hp"].mean()
print(f'{mean_3:.2f}')

167.28


### 5. Sort the `cars` DataFrame by `cyl` in ascending order, then `mpg` in descending order

In [64]:
cars.sort_values(["cyl","mpg"],ascending=[True,False]).head()

mpg  cyl   disp  hp  drat     wt   qsec  vs  am  gear  carb
19  33.9    4   71.1  65  4.22  1.835  19.90   1   1     4     1
17  32.4    4   78.7  66  4.08  2.200  19.47   1   1     4     1
18  30.4    4   75.7  52  4.93  1.615  18.52   1   1     4     2
25  27.3    4   79.0  66  4.08  1.935  18.90   1   1     4     1
7   24.4    4  146.7  62  3.69  3.190  20.00   1   0     4     2

In [65]:
cars.sort_values('cyl').sort_values('mpg',ascending=False).head()

mpg  cyl   disp  hp  drat     wt   qsec  vs  am  gear  carb
19  33.9    4   71.1  65  4.22  1.835  19.90   1   1     4     1
17  32.4    4   78.7  66  4.08  2.200  19.47   1   1     4     1
18  30.4    4   75.7  52  4.93  1.615  18.52   1   1     4     2
25  27.3    4   79.0  66  4.08  1.935  18.90   1   1     4     1
7   24.4    4  146.7  62  3.69  3.190  20.00   1   0     4     2

<a id='review'></a>

### Review 

We covered a lot of ground! It's ok if this takes a while to gel.

```python

# basic DataFrame operations
df.head()
df.tail()
df.shape
df.columns
df.Index
df.info()

# selecting columns
df.column_name
df['column_name']
df[['column_name1','column_name2','column_name3']]

# notable columns operations
df.describe() # five number summary
df['col1'].nunique() # number of unique values
df['col1'].value_counts() # number of occurrences of each value in column

# filtering
df[ df['col1'] < 50 ] # filter column to be less than 50
df[ (df['col1'] == value1) & (df['col2'] > value2) ] # filter column where col1 is equal to value1 AND col2 is greater to value 2

# sorting
df.sort_values(by='column_name', ascending = False) # sort biggest to smallest

```


It's common to refer back to your own code *all the time.* Don't hesistate to reference this guide! 🐼
